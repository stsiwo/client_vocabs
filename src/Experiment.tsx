import * as React from 'react';
import { withFormik, InjectedFormikProps } from 'formik';
import * as Yup from 'yup';

interface IDef {
  def: string;
}

interface IWord {
  name: string;
  defs: IDef[]; 
}

// words validation schema
//const WordsValidationSchema = Yup.object().shape({
  //words: Yup.array().of(Yup.object().shape({
    //name: Yup.string().required("required"),
    //defs: Yup.array().of(Yup.object().shape({
      //def: Yup.string().required("required"),
    //})),
  //})),
//});

const WordsValidationSchema = Yup.object().shape({
  words: Yup.array().of(Yup.object().shape({
    name: Yup.string().required("required"),
    defs: Yup.array().of(Yup.object().shape({
      def: Yup.string().required("required"),
    }))
  }))
})

//const arrayVAlidationSchema = Yup.object().shape({
  //myArray: Yup.array().of(Yup.object().shape({
    //name: Yup.string().required("required"),
  //})),
//});

//const validationSchema = Yup.object().shape({
  //name: Yup.string().required("required"),
//})

const testWords = [
  {
    name: "test1",
    defs: [
      {
        def: "def1",
      },
      {
        def: "def1",
      },
      {
        def: "def1",
      },
      {
        def: "def1",
      },
    ]
  },
  {
    name: "test2",
    defs: [
      {
        def: "def2",
      },
      {
        def: "def2",
      },
      {
        def: "def2",
      },
      {
        def: "def2",
      },
    ]
  },
]

//const testArray = [
  //{
    //name: "satoshi",
  //},
  //{
    //name: "hitomi",
  //},
  //{
    //name: "kaoru",
  //},
  //{
    //name: "yoich",
  //},
  //{
    //name: "seiko",
  //},
//]

interface Props  {
  words: IWord[];
  //name: string;
  //myArray: { name: string }[];
}

interface FormValues {
  words: IWord[];
  //name: string;
  //myArray: { name: string }[];
}

class Experiment extends React.Component<InjectedFormikProps<Props, FormValues>, {}> {
  constructor(props: InjectedFormikProps<Props, FormValues>) {
    super(props);
    this.renderWordForm = this.renderWordForm.bind(this);
  }

  renderWordForm() {

    //return this.props.values.myArray.map(( item, index ) => {
      //return (
        //<div key={ index }>
          //<input 
            //type="text" 
            //value={ item.name } 
            //name={ `myArray.${ index }.name` } 
            //onChange={ this.props.handleChange } 
            //onBlur={ this.props.handleBlur }
          ///>
        //</div>
      //)
    //});

    return this.props.values.words.map(( word, wordIndex ) => {
      return (
        <div key={ wordIndex }>
        {/*
          is there any clean way to display this error message???
          */}
          { this.props.errors.words && this.props.touched.words  && this.props.errors.words[wordIndex] && this.props.touched.words[wordIndex] && this.props.errors.words[wordIndex].name &&this.props.touched.words[wordIndex].name ? ( <div>{ this.props.errors.words[wordIndex].name} </div> ): null  } 
          <input 
            type="text" 
            value={ word.name } 
            name={ `words.${ wordIndex }.name` } 
            onChange={ this.props.handleChange } 
            onBlur={ this.props.handleBlur }
          />
          { word.defs.map(( def, defIndex ) => {
            return (
              <div key={ defIndex }>
              {/*
                { this.props.errors.words && this.props.errors.words[wordIndex].defs[defIndex].def && this.props.touched.words[wordIndex].defs[defIndex].def ? ( <div>this.props.errors.words[wordIndex].defs[defIndex].def</div> ): null  } 
                */}
                <input 
                  type="text" 
                  value={ def.def } 
                  name={ `words.${ wordIndex }.defs.${ defIndex }.def` } 
                  onChange={ this.props.handleChange } 
                  onBlur={ this.props.handleBlur }
                />
              </div>
            )
          })}
        </div>
      )
    })
  }

  render() {
    console.log(this.props);
    return (
      <form onSubmit={ this.props.handleSubmit } >
       
      {/*
        <input 
        type="text" 
        value={ this.props.values.name }
        onChange={ this.props.handleChange }
        onBlur={ this.props.handleBlur }
        name="name"
        />
        */}
        { this.renderWordForm() }
        <button type="submit" >submit</button>
      </form>
    );
  }
}

const FormExperience = withFormik<Props, FormValues>({
    // change props of this wrapping component into form values: set initial form values here
    mapPropsToValues: props => {
      return {
        words: props.words,
        //name: props.name,
        //myArray: props.myArray,
      };
    },

    // better to use validation schema (like Yup)
    // can be synchronous or asynchronous
    // return FormikErrors or Promise
    //validate: (values: FormValues) => {
      //let errors: FormikErrors<FormValues> = {};
      //return errors;
    //},

    // handle submit with submitted form values and extra object called "FormikBag"
    // FormikBag: includes props passed to wrapped component and some useful functions (see doc)
    handleSubmit: (values: FormValues) => {
      console.log(values); 
    },

    validationSchema: () => WordsValidationSchema,
})(Experiment);

const Form: React.SFC<{}> = () => {
  return <FormExperience words={ testWords } /*name="satoshi"*/ /*myArray={ testArray }*/ />
}
export default Form;

