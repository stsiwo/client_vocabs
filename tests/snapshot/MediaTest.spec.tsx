import * as React from 'react';
import Experiment from '../../src/Experiment';
import * as renderer from 'react-test-renderer';


test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <Experiment>test</Experiment> 
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
