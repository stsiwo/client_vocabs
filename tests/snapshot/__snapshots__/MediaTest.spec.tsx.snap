// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      name="words.0.name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="test1"
    />
    <div>
      <input
        name="words.0.defs.0.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def1"
      />
    </div>
    <div>
      <input
        name="words.0.defs.1.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def1"
      />
    </div>
    <div>
      <input
        name="words.0.defs.2.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def1"
      />
    </div>
    <div>
      <input
        name="words.0.defs.3.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def1"
      />
    </div>
  </div>
  <div>
    <input
      name="words.1.name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="test2"
    />
    <div>
      <input
        name="words.1.defs.0.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def2"
      />
    </div>
    <div>
      <input
        name="words.1.defs.1.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def2"
      />
    </div>
    <div>
      <input
        name="words.1.defs.2.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def2"
      />
    </div>
    <div>
      <input
        name="words.1.defs.3.def"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="def2"
      />
    </div>
  </div>
  <button
    type="submit"
  >
    submit
  </button>
</form>
`;
